README
README file for MarpaX::Languages::SVG::Parser.

See also: CHANGES and Changelog.ini.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip MarpaX-Languages-SVG-Parser-1.00.tgz
shell>tar mxvf MarpaX-Languages-SVG-Parser-1.00.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run Parser.pm through your favourite pod2html translator.

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip MarpaX-Languages-SVG-Parser-1.00.zip
shell>ppm install --location=. MarpaX-Languages-SVG-Parser
shell>del MarpaX-Languages-SVG-Parser-1.00.ppd
shell>del PPM-MarpaX-Languages-SVG-Parser-1.00.tar.gz


Changes
Revision history for Perl extension MarpaX::Languages::SVG::Parser.

1.05  Wed Sep 17 17:22:00 2014
	- Update Build.PL and Makefile.PL re putting this package on github.

1.04  Thu May 27 08:31:00 2014
	- Remove t/version.t, and so remove Test::Version, since the latter fails when combined with Test::Pretty.
		I do not use Test::Pretty, only Test::More.

1.03  Sat Feb 22 17:44:00 2014
	- Add 'lexeme default = latm => 1' to the grammar.
	- Make V 2.082 the minimum version for Marpa::R2.

1.02  Fri Nov  1 15:51:00 2013
	- Stop tests from requiring *::Utis, which uses *::Config, which is for my use only.
	- Simplify t/test.fake.data.t by not 'use'ing unused modules.

1.01  Tue Oct 29 16:23:00 2013
	- Set utf8 (sic) unilaterally on the log output, if logging. This allows STDOUT to be redirected.
		For non-utf8 input files (data/*.svg except data/utf8.01.svg), it will not be noticed.
		For data/utf8.01.svg, and similar user files, it is vital if log ouptut is being captured.
		Note: utf-8 (sic) is not supported by Log::Handler, as of V 0.80. I'll send the author a patch.
	- Fix counting of points within polygons and polylines. This was only a convenience, but was a bug.
	- Change hashref output in the case of commands, as in data/ellipse.02.log:
			Item  Type              Name              Value
			29    command           -                 M
		To what it should have been all along:
			29    command           M                 -
	- Eliminate the action object, and use some globals instead. This means the action class was re-written
		to avoid using Moo. Global variables are declared in the action class, and initialized in the SAX handler
		class. This is hardly ideal. We await Marpa::R3.
	- The action functions now return the hashrefs which used to be pushed onto a stack.
	- Add the decode_result() method in SAXHandler.pm, to handle arbitrarily nested arrayrefs and hashrefs.
		This is the code required to unscramble what's returned by Marpa as the result of the parse.
		The unscrambling makes the hashrefs returned by the action subs available as a list.
	- Edit grammar slightly because we now use Marpa instead of Set::Array to gather results.
		This just consists in making some tokens hidden from the semantics, i.e. in changing:
			translate ::= 'translate'  '('  number_set  ')'  action => command
		To:
			translate ::= 'translate' ('(') number_set (')') action => command
	- Update docs in each module.
	- Add scripts/svg2log.pl to convert data/*.svg into data/*.log.
	- Add t/test.real.data.t to check output of parse.file.pl against data/*.log.
	- Rename t/runner.t to t/test.fake.data.t.

1.00  Thu Oct 20 13:34:00 2013
	  - Original version


Build.PL
use strict;
use warnings;

use Module::Build;

require 5.014002; # For the utf8 stuff.

# -----------------------------------------------

my $class = Module::Build->subclass(code => <<'EOF');

sub ACTION_authortest
{
	my($self) = @_;

	$self->depends_on('build');
	$self->depends_on('manifest');
	$self->depends_on('distmeta');

	$self->test_files( qw< t xt/author > );
	$self->recursive_test_files(1);

	$self->depends_on('test');

	return;
}
EOF

$class -> new
(
	module_name    => 'MarpaX::Languages::SVG::Parser',
	license        => 'artistic_2',
	dist_abstract  => 'An SVG parser',
	dist_author    => 'Ron Savage <ron@savage.net.au>',
	build_requires =>
	{
		'Test::More'    => 0.94,
# 	 	'Test::Pod'     => 1.45, # Make it optional. See xt/author/pod.t
#		'Test::Version' => 1.002003,
	},
	configure_requires =>
	{
		'Module::Build' => 0.3800,
	},
	meta_merge =>
	{
		resources =>
		{
			repository => 'https://github.com/dwimperl/MarpaX-Languages-SVG-Parser',
		},
	},
	requires =>
	{
		'Capture::Tiny'           => 0.22,
#		'Config'                  => undef, # Version is unparsable.
		'Config::Tiny'            => 2.19,
		'Data::Section::Simple'   => 0.05,
		'Date::Simple'            => 3.03,
		'File::Basename'          => 2.82,
		'File::Copy'              => 2.21,
		'File::HomeDir'           => 1.00,
		'File::Slurp'             => 9999.19,
		'File::Spec'              => 3.40,
		'Getopt::Long'            => 2.41,
		'Log::Handler'            => 0.79,
		'Marpa::R2'               => 2.082000,
		'Moo'                     => 1.003001,
		'open'                    => 0,
		'Path::Tiny'              => 0.034,
#		'perl'                    => undef, # For the utf8 stuff. # Version is unparsable.
		'Pod::Usage'              => 1.63,
		'Set::Array'              => 0.30,
		'strict'                  => 0,
		'Text::CSV::Encoded'      => 0.22,
		'Text::Xslate'            => 1.6002,
		'Types::Standard'         => 0.026,
		'utf8'                    => 0,
		'warnings'                => 0,
		'XML::SAX::Base'          => 1.08,
		'XML::SAX::ParserFactory' => 1.01,
	},
	resources =>
	{
		bugtracker => 'https://rt.cpan.org/Public/Dist/Display.html?Name=MarpaX-Languages-SVG-Parser',
		license    => 'http://www.perlfoundation.org/artistic_license_2_0',
	},
) -> create_build_script();


scripts/parse.file.pl
#!/usr/bin/env perl

use strict;
use utf8;
use warnings;
use open qw(:std :utf8); # Undeclared streams in UTF-8.

use Getopt::Long;

use MarpaX::Languages::SVG::Parser;

use Pod::Usage;

# -----------------------------------------------

my($option_parser) = Getopt::Long::Parser -> new();

my(%option);

if ($option_parser -> getoptions
(
	\%option,
	'encoding=s',
	'help',
	'input_file_name=s',
	'maxlevel=s',
	'minlevel=s',
	'output_file_name=s',
) )
{
	pod2usage(1) if ($option{'help'});

	exit MarpaX::Languages::SVG::Parser -> new(%option) -> run;
}
else
{
	pod2usage(2);
}

__END__

=pod

=head1 NAME

parse.file.pl - Parsing any SVG file

=head1 SYNOPSIS

test.file.pl [options]

	Options:
	-encoding aString
	-help
	-input_file_name anSVGFileName
	-maxlevel aString
	-minlevel aString
	-output_file_name aCSVFileName

All switches can be reduced to a single letter.

Exit value: 0.

=head1 OPTIONS

=over 4

=item o -encoding aString

aString takes values such as 'utf-8', and the code converts this into '<:encoding(utf-8)'.

This option is rarely needed for input, but may be for output.

Default: ''.

=item o -help

Print help and exit.

=item o -input_file_name anSVGFileName

The name of an SVG file to process.

See data/*.svg for some samples.

This option is mandatory.

Default: ''.

=item o -maxlevel logOption1

This option affects Log::Handler.

See the Log::handler docs.

Default: 'notice'.

=item o -minlevel logOption2

This option affects Log::Handler.

See the Log::handler docs.

Default: 'error'.

No lower levels are used.

=item o -output_file_name aCSVFileName

The name of a CSV file to write, of parsed tokens.

See also the -encoding option.

By default, nothing is written.

Default: ''.

=back

=cut


lib/MarpaX/Languages/SVG/Parser/Utils.pm
package MarpaX::Languages::SVG::Parser::Utils;

use strict;
use utf8;
use warnings;
use warnings  qw(FATAL utf8);    # Fatalize encoding glitches.

use Config;

use Date::Simple;

use File::Basename; # For basename().
use File::Spec;

use MarpaX::Languages::SVG::Parser::Config;

use Moo;

use Text::Xslate 'mark_raw';

use Types::Standard qw/HashRef/;

has config =>
(
	default  => sub{return MarpaX::Languages::SVG::Parser::Config -> new -> config},
	is       => 'rw',
	isa      => HashRef,
	required => 0,
);

our $VERSION = '1.05';

# ------------------------------------------------

sub generate_demo
{
	my($self) = @_;

	# Generate html/*.svg.

	`$^X -Ilib scripts/bnf2graph.pl`;

	my($basename);
	my(%data_file);
	my($image);

	for my $file (MarpaX::Languages::SVG::Parser::Utils -> new -> get_files('data', 'bnf') )
	{
		$basename             = basename($file);
		$image                = $basename =~ s/bnf$/svg/r;
		$data_file{$basename} =
		{
			bnf   => $file,
			image => $image,
		};
	}

	my($config)    = $self -> config;
	my($templater) = Text::Xslate -> new
	(
		input_layer => '',
		path        => $$config{template_path},
	);
	my($count) = 0;
	my($index) = $templater -> render
	(
	'marpax.languages.svg.parser.tx',
	{
		default_css => "$$config{css_url}/default.css",
		data        =>
			[
			map
			{
				{
					bnf    => mark_raw($data_file{$_}{bnf}),
					count  => ++$count,
					image  => $data_file{$_}{image},
				};
			} sort keys %data_file
			],
		environment     => $self -> generate_demo_environment,
		fancy_table_css => "$$config{css_url}/fancy.table.css",
		version         => $VERSION,
	}
	);
	my($file_name) = File::Spec -> catfile('html', 'index.html');

	open(OUT, '>', $file_name);
	print OUT $index;
	close OUT;

	print "Wrote $file_name\n";

	# Return 0 for success and 1 for failure.

	return 0;

} # End of generate_demo.

# ------------------------------------------------

sub generate_demo_environment
{
	my($self) = @_;

	my(@environment);

	# mark_raw() is needed because of the HTML tag <a>.

	push @environment,
	{left => 'Author', right => mark_raw(qq|<a href="http://savage.net.au/">Ron Savage</a>|)},
	{left => 'Date',   right => Date::Simple -> today},
	{left => 'OS',     right => 'Debian V 6'},
	{left => 'Perl',   right => $Config{version} };

	return \@environment;

} # End of generate_demo_environment.

# ------------------------------------------------

1;

=pod

=head1 NAME

C<MarpaX::Languages::SVG::Parser::Utils> - A nested SVG parser, using XML::SAX and Marpa::R2

=head1 Synopsis

See L<MarpaX::Languages::SVG::Parser/Synopsis>.

=head1 Description

Utility routines for L<MarpaX::Languages::SVG::Parser>.

Only used by scripts/*.pl and t/*.t.

=head1 Installation

See L<MarpaX::Languages::SVG::Parser/Installation>.

=head1 Constructor and Initialization

See L<MarpaX::Languages::SVG::Parser/Constructor and Initialization>.

=head1 Methods

=head2 generate_demo_environment()

Returns a hashref of OS, etc, values.

Keys are C<left> and C<right>, to suit C<htdocs/assets/templates/marpax/languages/svg/parser/fancy.table.tx>.

C<*.tx> files are used by L<Text::Xslate>.

Called by L</generate_demo_index()>.

=head2 generate_demo_index()

Calls L<MarpaX::Languages::SVG::Parser::Utils/get_files($dir_name, $type)> and L</generate_demo_environment()>.

Writes C<html/index.html>.

See scripts/generate.demo.pl.

=head2 new()

This method is auto-generated by L<Moo>.

=head1 Author

L<MarpaX::Languages::SVG::Parser> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2013.

Home page: L<http://savage.net.au/>.

=head1 Copyright

Australian copyright (c) 2013, Ron Savage.

	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License 2.0, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut


lib/MarpaX/Languages/SVG/Parser/Config.pm
package MarpaX::Languages::SVG::Parser::Config;

use strict;
use utf8;
use warnings;
use warnings  qw(FATAL utf8);    # Fatalize encoding glitches.

use Config::Tiny;

use File::HomeDir;

use Moo;

use Path::Tiny; # For path().

use Types::Standard qw/Any Str/;

has config =>
(
	default  => sub{return {} },
	is       => 'rw',
	isa      => Any, # Cannot use HashRef. See line 65.
	required => 0,
);

has config_file_path =>
(
	default  => sub{return ''},
	is       => 'rw',
	isa      => Any, # Can't use Str. See line 61.
	required => 0,
);

has section =>
(
	default  => sub{return ''},
	is       => 'rw',
	isa      => Str,
	required => 0,
);

our $VERSION = '1.05';

# -----------------------------------------------

sub BUILD
{
	my($self) = @_;
	my($path) = path(File::HomeDir -> my_dist_config('MarpaX-Languages-SVG-Parser'), '.htmarpax.languages.svg.parser.conf');

	$self -> read($path);

} # End of BUILD.

# -----------------------------------------------

sub read
{
	my($self, $path) = @_;

	$self -> config_file_path($path);

	# Check [global].

	$self -> config(Config::Tiny -> read($path) );

	if (Config::Tiny -> errstr)
	{
		die Config::Tiny -> errstr;
	}

	$self -> section('global');

	if (! ${$self -> config}{$self -> section})
	{
		die "Config file '$path' does not contain the section [@{[$self -> section]}]\n";
	}

	# Check [x] where x is host=x within [global].

	$self -> section(${$self -> config}{$self -> section}{'host'});

	if (! ${$self -> config}{$self -> section})
	{
		die "Config file '$path' does not contain the section [@{[$self -> section]}]\n";
	}

	# Move desired section into config, so caller can just use $self -> config to get a hashref.

	$self -> config(${$self -> config}{$self -> section});

}	# End of read.

# --------------------------------------------------

1;

=pod

=head1 NAME

MarpaX::Languages::SVG::Parser::Config - A config manager for use by MarpaX::Languages::SVG::Parser

=head1 Synopsis

See L<MarpaX::Languages::SVG::Parser/Synopsis>.

=head1 Description

Basically just utility routines for L<MarpaX::Languages::SVG::Parser>.

Only used to generate the SVGs of the BNFs.

=head1 Installation

See L<MarpaX::Languages::SVG::Parser/Installation>.

=head1 Methods

=head2 read()

read() is called automatically by new(). It does the actual reading of the config file.

If the file can't be read, C<die $string> is called.

The path to the config file is determined by:

	path(File::HomeDir -> my_dist_config('MarpaX-Languages-SVG-Parser'), '.htmarpax.languages.svg.parser.conf');

The author runs scripts/copy.config.pl, which uses similar code, to move the config file
from the config/ directory in the disto into an OS-dependent directory.

The run-time code uses this module to look in the same directory as used by scripts/copy.config.pl.

=head1 Support

Email the author, or log a bug on RT:

L<https://rt.cpan.org/Public/Dist/Display.html?Name=MarpaX::Languages::SVG::Parser>.

=head1 Author

L<MarpaX::Languages::SVG::Parser> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2013.

Home page: L<http://savage.net.au/index.html>.

=head1 Copyright

Australian copyright (c) 2013, Ron Savage.

	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut


lib/MarpaX/Languages/SVG/Parser/Actions.pm
package MarpaX::Languages::SVG::Parser::Actions;

use strict;
use utf8;
use warnings;
use warnings  qw(FATAL utf8);    # Fatalize encoding glitches.

my($item_count);
my($param_count);

our $logger;

our $VERSION = '1.05';

# ------------------------------------------------

sub boolean
{
	my($hashref, $t1) = @_;
	$t1 = lc $t1;
	$t1 = $t1 eq 'zero' ? 0 : 1;

	return
	{
		count => ++$item_count,
		name  => ++$param_count,
		type  => 'boolean',
		value => $t1,
	};

} # End of boolean.

# ------------------------------------------------

sub command
{
	my($hashref, $t1, @t2) = @_;
	$param_count = 0;

	return
	{
		count => ++$item_count,
		name  => $t1,
		type  => 'command',
		value => [@t2],
	};

} # End of command.

# ------------------------------------------------

sub float
{
	my($hashref, $t1) = @_;

	return
	{
		count => ++$item_count,
		name  => ++$param_count,
		type  => 'float',
		value => $t1,
	};

} # End of float.

# ------------------------------------------------

sub init
{
	$item_count  = 0;
	$param_count = 0;

} # End of init.

# ------------------------------------------------

sub integer
{
	my($hashref, $t1) = @_;

	return
	{
		count => ++$item_count,
		name  => ++$param_count,
		type  => 'integer',
		value => $t1,
	};

} # End of integer.

# --------------------------------------------------

sub log
{
	my($level, $s) = @_;
	$level = 'notice' if (! defined $level);
	$s     = ''       if (! defined $s);

	$logger -> $level($s) if ($logger);

} # End of log.

# ------------------------------------------------

sub string
{
	my($hashref, $t1) = @_;

	return
	{
		count => ++$item_count,
		name  => ++$param_count,
		type  => 'string',
		value => $t1,
	};

} # End of string.

# ------------------------------------------------

1;

=pod

=head1 NAME

C<MarpaX::Languages::SVG::Parser::Actions> - A nested SVG parser, using XML::SAX and Marpa::R2

=head1 Synopsis

See L<MarpaX::Languages::SVG::Parser/Synopsis>.

=head1 Description

Basically just utility routines for L<MarpaX::Languages::SVG::Parser>. Only used indirectly by L<Marpa::R2>.

Specifially, calls to functions are triggered by items in the input stream matching elements of the current
grammar (and Marpa does the calling).

Each action function returns a hashref, which Marpa gathers. The calling code
L<MarpaX::Languages::SVG::Parser::SAXHandler> decodes the result and puts the hashrefs into a stack, described in
the L<MarpaX::Languages::SVG::Parser/FAQ>.

=head1 Installation

See L<MarpaX::Languages::SVG::Parser/Installation>.

=head1 Constructor and Initialization

This class has no constructor. L<Marpa::R2> fabricates an instance, but won't let us get access to it.

So, we use a global variable, C<$logger>, initialized in L<MarpaX::Languages::SVG::Parser::SAXHandler>,
in case we need logging. Details:

=over 4

=item o logger => aLog::HandlerObject

By default, an object of type L<Log::Handler> is created which prints to STDOUT,
but given the default, nothing is actually printed unless the C<maxlevel> attribute of this object is changed
in L<MarpaX::Languages::SVG::Parser>.

Default: anObjectOfTypeLogHandler.

=back

Also, each new parse is preceeded by a call to the L</init()> function, to reset some counters global to this file.

=head1 Methods

None.

=head1 Functions

=head2 boolean($t1)

Returns a hashref identifying the boolean $t1.

=head2 command($t1, @t2)

Returns a hashref identifying the command $t1 and its parameters in @t2.

=head2 float($t1)

Returns a hashref identifying the float $t1.

=head2 init()

Resets some counters global to the file. This must be called at the start of each new parse.

=head2 integer($t1)

Returns a hashref identifying the integer $t1.

=head2 log($level, $s)

Calls $logger -> log($level => $s) if ($logger).

=head2 string($t1)

Returns a hashref identifying the string $t1.

=head1 FAQ

See L<MarpaX::Languages::SVG::Parser/FAQ>.

=head1 Author

L<MarpaX::Languages::SVG::Parser> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2013.

Home page: L<http://savage.net.au/>.

=head1 Copyright

Australian copyright (c) 2013, Ron Savage.

	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License 2.0, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut


Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker;

require 5.014002; # For the utf8 stuff.

# -----------------------------------------------

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my(%params) =
(
	($] ge '5.005') ?
	(
		AUTHOR   => 'Ron Savage (ron@savage.net.au)',
		ABSTRACT => 'An SVG parser',
	) : (),
	clean =>
	{
		FILES => 'blib/* Makefile MANIFEST MarpaX-Languages-SVG-Parser*'
	},
	dist =>
	{
		COMPRESS => 'gzip',
		SUFFIX   => 'gz'
	},
	DISTNAME  => 'MarpaX-Languages-SVG-Parser',
	NAME      => 'MarpaX::Languages::SVG::Parser',
	PL_FILES  => {},
	PREREQ_PM =>
	{
		'Capture::Tiny'           => 0.22,
#		'Config'                  => undef, # Version is unparsable.
		'Config::Tiny'            => 2.19,
		'Data::Section::Simple'   => 0.05,
		'Date::Simple'            => 3.03,
		'File::Basename'          => 2.82,
		'File::Copy'              => 2.21,
		'File::HomeDir'           => 1.00,
		'File::Slurp'             => 9999.19,
		'File::Spec'              => 3.40,
		'Getopt::Long'            => 2.41,
		'Log::Handler'            => 0.79,
		'Marpa::R2'               => 2.082000,
		'Moo'                     => 1.003001,
		'open'                    => 0,
		'Path::Tiny'              => 0.034,
#		'perl'                    => undef, # For the utf8 stuff. # Version is unparsable.
		'Pod::Usage'              => 1.63,
		'Set::Array'              => 0.30,
		'strict'                  => 0,
		'Test::More'              => 0.94,
#	 	'Test::Pod'               => 1.45, # Make it optional. See xt/author/pod.t
#		'Test::Version'           => 1.002003,
		'Text::CSV::Encoded'      => 0.22,
		'Text::Xslate'            => 1.6002,
		'Types::Standard'         => 0.026,
		'utf8'                    => 0,
		'warnings'                => 0,
		'XML::SAX::Base'          => 1.08,
		'XML::SAX::ParserFactory' => 1.01,
	},
	VERSION_FROM => 'lib/MarpaX/Languages/SVG/Parser.pm',
);

if ( ($ExtUtils::MakeMaker::VERSION =~ /^\d\.\d\d$/) && ($ExtUtils::MakeMaker::VERSION > 6.30) )
{
	$params{LICENSE} = 'artistic_2';
}

if ($ExtUtils::MakeMaker::VERSION ge '6.46')
{
	$params{META_MERGE} =
	{
		resources =>
		{
			bugtracker => 'https://rt.cpan.org/Public/Dist/Display.html?Name=MarpaX-Languages-SVG-Parser',
			license    => 'http://www.perlfoundation.org/artistic_license_2_0',
		},
	};
}

WriteMakefile(%params);
